### DISCUSSION TITLE
Obtain the wind speed at each turbine after wake effect

### DISCUSSION THREAD

**Discussion by cjeong3** [NatLabRockies/floris#802] [Q&A] [answered]
Hi!

I have several questions about using FLROIS.
I have a (100x4) matrix-type meteorological (or atmospheric) data right before entering a wind farm, consisting of 
- columns: incoming wind speed (WS), wind direction (WD), turbulence intensity (TI), and air density (AD);
- rows: 100 different combinations of 4 variables.

This wind farm has 20 wind turbines. 

The questions are:

Q1. Can I get the **waked** wind speed at each turbine (right before each rotor, i.e., before applying the power curve at each turbine) rather than power? It seems we can only get powers at each turbine from `get_turbin_powers()` after excuting `calculate_wake()`. But, in my analysis, it is important to obtain the value of wind deficits at each turbine from the wake effect. If not, is there any way to get the waked wind speed at each turbine?

Q2. I'd like to calculate waked wind speed at each turbine (like Q1) for each combination of 4 variables (WS, WD, TI, AD) with 100 observations by executing **one** FLORIS evaluation `calculate_wake()`. But, the instruction indicates that, for example, 2 WS and 2 WD in one 'yaml' input file simulate total 4 results, and only one TI can be configured in the 'yaml' file for one simulation. Does it mean I should simulate FLORIS 100 times for each combination? (100 times is fine now but what if I consider 10000 observations?) Are there other ways to do the same task efficiently with minimal FLORIS evaluation? 

Q3. Does several wake models used in FLORIS consider the wake effect in the multi-turbine wind farm setting? The reason why I ask is that the turbines affect others and are affected by others complicatedly in the multi-turbine wind farm. As far as I know, the wakes from a single turbine and multi-turbines should behave differently. 

Thank you in advance!

**Comment by rafmudaf** [comment, marked as answer]
1. Yes, FLORIS calculates the incoming velocities at the rotor and then uses that to calculate the thrust and power. You can access them through `FlorisInterface.floris.flow_field.u_sorted`. If you need the associated locations, you can get those with `FlorisInterface.floris.grid.x_sorted` and similar for y and z. See https://github.com/NREL/floris/discussions/738#discussioncomment-8387488 for a little more explanation. Note that this is the waked wind speed directly, not just the velocity deficit.
2. Currently, FLORIS supports the wind speed and wind direction product option and a time series option where a list of wind speed and wind direction combinations is given. The second option is probably what you want, and you would still have to add a loop to your analysis script to vary TI and AD. This is enabled through `FlorisInterface.reinitialize(wind_directions=..., wind_speeds=..., time_series=True)`. See [this example](https://github.com/NREL/floris/blob/main/examples/21_demo_time_series.py) for more information. Also heads up that we are currently working on the next major release, FLORIS v4, that supports this type of input directly. It will be available at the end of March. You can review the active development in the [v4 project](https://github.com/orgs/NREL/projects/96).
3. It sounds like you are asking about the wake combination models. FLORIS includes three wake combination models, and you can see a comparison of them [in the docs](https://nrel.github.io/floris/wake_models.html#wake-combination-models). Please let me know if this is not what you're referring to.

**Reply by cjeong3** [reply]
Thank you for the detailed answers, @rafmudaf
It really helps a lot!
I'll test it as you suggested.

**Reply by Nellytado** [reply]
Hi @rafmudaf, 

How do I exactly use FlorisInterface.floris.flow_field.u_sorted in my code to extract the wind speeds at the turbines? After I have loaded the gch file and imported all the necessary floris interfaces. 

Thanks.

**Reply by paulf81** [reply]
HI @Nellytado , have a look at this example: https://github.com/NREL/floris/blob/main/examples/22_get_wind_speed_at_turbines.py
